### PR TITLE
Ai script

### PR DESCRIPTION
**Pull request by GitBluub**
## Proposed changes
AIControllable using lua

IAControllableComponent and System
Component takes a filepath in ctor to the lua script

System updates internal map and register a set of functions for the lua script to use

## Information
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Non-breaking changes

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] I have added necessary documentation

### CODE REVIEW

File: sources/Map/LuaMap.cpp [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+		std::vector<Vector2f> path;
+
+		open.push_back(root);
+		gScore[root] = 0;
+		fScore[root] = root.distance(target);
+
+		while (open.size())
+		{
+			auto min_elem = std::min_element(open.begin(), open.end(),
+			[&fScore](const decltype(open)::value_type &l, const decltype(open)::value_type &r) -> bool
```
Comment (AnonymusRaccoon) [inline sources/Map/LuaMap.cpp:102]: use const auto &, you don't need a declytype here.

File: sources/Map/LuaMap.cpp [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+	Vector2f LuaMap::findSafeSpace(void) const
+	{
+		return _roundedPlayer;
+	}
+
+	int LuaMap::getMap(lua_State *L)
+	{
+		LuaG::State state(L);
+		int index = 1;
+    	const LuaMap *map = (const LuaMap *) state.getPointer(state.getFirstUpValueIdx());
```
Comment (AnonymusRaccoon) [inline sources/Map/LuaMap.cpp:149]: Don't use C-cast, I believe you want to reinterpret_cast here.

File: sources/Map/LuaMap.cpp [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+			state.setTable();
+		}
+		return 1;	
+	}
+	
+	int LuaMap::getDanger(lua_State *L)
+	{
+		LuaG::State state(L);
+		int index = 1;
+    	const LuaMap *map = (const LuaMap *) state.getPointer(state.getFirstUpValueIdx());
```
Comment (AnonymusRaccoon) [inline sources/Map/LuaMap.cpp:169]: C-cast

File: sources/Map/LuaMap.hpp [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+namespace std
+{
+	template<>
+	struct hash<BBM::Vector2f>
+	{
+		typedef BBM::Vector2f argument_type;
+		typedef std::size_t result_type;
+		result_type operator()(argument_type const &in) const
+		{
+			return in.x + in.y;
```
Comment (AnonymusRaccoon) [inline sources/Map/LuaMap.hpp:97]: This is not hashing. You could hash the vector like that:
```c++
return in.x | (in.y << 32);
```

File: sources/Map/Map.cpp [hunk 5]
```diff
@@ -240,9 +240,9 @@ namespace BBM
 			.addComponent<TagComponent<Blowable>>()
 			.addComponent<TagComponent<Breakable>>()
 			.addComponent<HealthComponent>(1, &MapGenerator::wallDestroyed)
-			.addComponent<CollisionComponent>(
```
Comment (AnonymusRaccoon) [inline sources/Map/Map.cpp:243]: Don't remove collisions in a PR

Comment (GitBluub) [reply sources/Map/Map.cpp:243]: This is still a draft

File: sources/Map/LuaMap.cpp (hunk 2 above)
Comment (GitBluub) [reply sources/Map/LuaMap.cpp]: fixed -> 638c134

File: sources/Map/LuaMap.cpp (hunk 1 above)
Comment (GitBluub) [reply sources/Map/LuaMap.cpp]: fixed -> 0d49984

**Review by AnonymusRaccoon — APPROVED** [review]

File: sources/Map/LuaMap.hpp (hunk 4 above)
Comment (AnonymusRaccoon) [reply sources/Map/LuaMap.hpp:104]:
```suggestion
			union {
				float vector[2];
				result_type hashed;
			} hasher;
			hasher.vector[0] = in.x;
			hasher.vector[1] = in.y;
			return hasher.hashed;
```

**Review by Octopus773 — APPROVED** [review]

File: sources/Map/LuaMap.hpp (hunk 4 above)
Comment (GitBluub) [reply sources/Map/LuaMap.hpp]: fixed -> 6342c6b

**Pull request merged**